File: Graph Data Structures/Weighted-Graphs.py
class PriorityQueue(object):
    def __init__(self):
        self.queue = list()
    
    def add(self, Edge):
        self.queue.append(Edge)
        for i in range(0, len(self.queue)):
                already_sorted = True
                for j in range(len(self.queue) - i - 1):
                    if (self.queue[j].weight > self.queue[j+1].weight):
                        self.queue[j], self.queue[j+1] = self.queue[j+1], self.queue[j]
                    already_sorted = False
                if already_sorted:
                    break
    def pop(self):
        return self.queue.pop(0)
    
    def isEmpty(self):
        return not self.queue

class Edge(object):
    def __init__(self, source, connection, weight):
        self.source = source
        self.connection = connection
        self.weight = weight

class WeightedGraph(object):
    def __init__(self, V):
        self.V = V
        self.adj = list()
        for i in range(0, self.V):
            x = list()
            self.adj.append(x)
    
    def addEdge(self, v, w, weight):
        self.adj[v].append(Edge(v, w, weight))
        self.adj[w].append(Edge(w, v, weight))
    
    def returnAdjacent(self, v):
        return self.adj[v]
    
    def printG(self):
        for i in range(0, self.V):
            print(str(i), end=" | ")
            for j in range(0, len(self.adj[i])):
                print(str(self.adj[i][j].connection) + " = " + str(self.adj[i][j].weight), end=", ")
            print()
    def getV(self):
        return self.V

class Dijkstra(object):
    def __init__(self, graph, s):
        self.s = s
        self.pq = PriorityQueue()
        self.edgeTo = list()
        self.edgeTo = [None] * graph.getV()
        self.distTo = list()
        self.distTo = [9999] * graph.getV()
        self.marked = list()
        for i in range (0, graph.getV()):
            if (i != self.s):
                self.distTo[i] = 9999
        self.distTo[s] = 0
        self.pq.add(Edge(s, s, 0))
        self.dijkstra(graph, s)
    
    def dijkstra(self, graph, v):
        while (self.pq.isEmpty() == False):
            n = self.pq.pop()
            self.marked.append(n.connection)
            for i in graph.returnAdjacent(n.connection):
                if (i.connection not in self.marked):
                    if (self.distTo[n.connection] + i.weight < self.distTo[i.connection]):
                        self.distTo[i.connection] = self.distTo[n.connection] + i.weight
                        self.edgeTo[i.connection] = n.connection
                        self.pq.add(Edge(n.connection, i.connection, self.distTo[i.connection]))
    
    def pathTo(self, t):
        path = list()
        if (self.hasPathTo(t) == False):
            return path
        i = t
        while (i != self.s):
            path.append(i)
            i = self.edgeTo[i]
        path.append(self.s)
        return path[::-1]

        
    def hasPathTo(self, t):
        if (t in self.marked):
            return True
        return False
    
    def getDistTo(self):
        return self.distTo
    
    def getMarked(self):
        return self.marked
    
    def getEdgeTo(self):
        return self.edgeTo

class WeightedQuickUnion(object):
    def __init__(self, size):
        self.sets = list()
        for i in range (0, size):
            self.sets.append(-1)
    
    def root(self, current):
        i = current
        while (self.sets[i] != -1):
            i = self.sets[i]
        return i
    
    def isConnected(self, s, t):
        return self.root(s) == self.root(t)
    
    def connect(self, s, t):
        self.sets[self.root(s)] = self.root(t)

class Kruskal(object):
    def __init__(self, graph):
        self.pq = PriorityQueue()
        self.mst = list()
        for i in range(0, graph.getV()):
            for j in graph.returnAdjacent(i):
                self.pq.add(j)
        
        self.kruskal(graph)

    def kruskal(self, graph):
        wqu = WeightedQuickUnion(graph.getV())
        while (self.pq.isEmpty() == False):
            n = self.pq.pop()
            if (wqu.isConnected(n.source, n.connection) == False):
                wqu.connect(n.source, n.connection)
                self.mst.append(n)
    
    def returnMST(self):
        return self.mst

wg = WeightedGraph(11)
wg.addEdge(0, 1, 1)
wg.addEdge(1, 2, 2)
wg.addEdge(1, 4, 3)
wg.addEdge(2, 5, 4)
wg.addEdge(3, 4, 5)
wg.addEdge(4, 5, 6)
wg.addEdge(5, 6, 7)
wg.addEdge(5, 8, 8)
wg.addEdge(6, 7, 9)
wg.printG()
print("---")
start = 0
end = 7
d = Dijkstra(wg, start)

print("Shortest Path From " + str(start) + " to " + str(end) + ": ")

for i in range(0, len(d.pathTo(end))):
    if (i == len(d.pathTo(end))-1):
        print(d.pathTo(end)[i])
    else:
        print(d.pathTo(end)[i], end=" => ")
l = d.getDistTo()
print("Total Weight: " + str(l[end]))
print("---")



k = Kruskal(wg)
print("MST: ")
for i in k.returnMST():
    print(str(i.source) + " <=> " + str(i.connection) + ": " + str(i.weight))

File: Graph Data Structures/Weighted-Graphs.cpp
//Edge.hpp
#include <iostream>
#include <vector>
using namespace std;

struct Edge{
    public:
        int source;
        int connection;
        int weight;
        Edge(int s, int c, int w);
};

//Edge.cpp
#include "Edge.hpp"
#include <iostream>
#include <vector>
using namespace std;

Edge::Edge(int s, int c, int w){
    source = s;
    connection = c;
    weight = w;
}

//Compare.cpp
#include "Edge.cpp"
class Compare{
    public:
        bool operator()(const Edge* a, const Edge* b) const{
            return a->weight > b->weight;
        }
};

//WeightedGraph.hpp
#include <iostream>
#include <vector>
#include "Compare.cpp"
using namespace std;

class WeightedGraph{
    public:
        int V;
        std::vector<std::vector<Edge>> adj;
        WeightedGraph(int v);
        void addEdge(int v, int w, int weight);
        std::vector<Edge> returnAdjacent(int v);
        void print();
};

//WeightedGraph.cpp
#include <iostream>
#include <vector>
#include "WeightedGraph.hpp"
using namespace std;

WeightedGraph::WeightedGraph(int v){
    V = v;
    std::vector<vector<Edge>> temp(V);
    adj = temp;
    for (int i = 0; i < V; i++){
        std::vector<Edge> neighbors;
        adj[i] = neighbors;
    }
}

void WeightedGraph::addEdge(int v, int w, int weight){
    adj[v].push_back(Edge(v, w, weight));
    adj[w].push_back(Edge(w, v, weight));
}

std::vector<Edge> WeightedGraph::returnAdjacent(int v){
    return adj[v];
}

void WeightedGraph::print(){
    for (int i = 0; i < V; i++){
        std::cout << i << " | ";
        for (int j = 0; j < adj[i].size(); j++){
            std::cout << adj[i][j].connection << " = " << adj[i][j].weight << ", ";
        }
        std::cout << "\n";
    }
}

//Dijkstra.hpp
#include <iostream>
#include <vector>
#include "WeightedGraph.cpp"
#include <queue>
using namespace std;
class Dijkstra{
    public:
        std::priority_queue<Edge*, std::vector<Edge*>, Compare> pq;
        std::vector<int> edgeTo;
        std::vector<int> distTo;
        std::vector<int> marked;
        int S;
        int V;
        Dijkstra(WeightedGraph g, int s);
        void dijkstra(WeightedGraph g, int s);
        std::vector<int> pathTo(int t);
        bool hasPathTo(int t);
        std::vector<int> getDistTo(int t);
};

//Dijkstra.cpp
#include "Dijkstra.hpp"
#include <iostream>
#include <vector>
#include <queue>
#include <limits.h>
#include <algorithm>
using namespace std;

Dijkstra::Dijkstra(WeightedGraph g, int s){
    V = g.V;
    std::vector<int> tempEdgeTo(V);
    edgeTo = tempEdgeTo;
    std::vector<int> tempDistTo(V);
    distTo = tempDistTo;
    std::vector<int> tempMarked(V);
    marked = tempMarked;
    S = s;
    pq.push(new Edge(s, s, 0));
    distTo[s] = 0;
    for (int i = 0; i < V; i++){
        if (i != S){
            distTo[i] = INT_MAX;
        }
    }
    dijkstra(g, s);
}

void Dijkstra::dijkstra(WeightedGraph g, int s){
    while(!pq.empty()){
        Edge* n = pq.top();
        pq.pop();
        marked.push_back(n->connection);
        std::vector<Edge> neighbors = g.returnAdjacent(n->connection);
        for (Edge i : neighbors){
            if (std::find(marked.begin(), marked.end(), i.connection) == marked.end()){
                if (distTo[n->connection] + i.weight < distTo[i.connection]){
                    distTo[i.connection] = distTo[n->connection] + i.weight;
                    edgeTo[i.connection] = n->connection;
                    pq.push(new Edge(n->connection, i.connection, distTo[i.connection]));
                }
            }
        }
    }
}

std::vector<int> Dijkstra::pathTo(int t){
    std::vector<int> path;
    if (hasPathTo(t) == false){
        return path;
    }
    for (int i = t; i != S; i = edgeTo[i]){
        path.push_back(i);
    }
    path.push_back(S);
    std::reverse(path.begin(), path.end());
    return path;
}

bool Dijkstra::hasPathTo(int t){
    if (std::find(marked.begin(), marked.end(), t) != marked.end()){
        return true;
    }
    return false;
}

std::vector<int> Dijkstra::getDistTo(int t){
    return distTo;
}

//WeightedQuickUnion.cpp
#include <iostream>
#include <vector>
using namespace std;
class WeightedQuickUnion{
    public:
        std::vector<int> sets;
        WeightedQuickUnion(int size){
            std::vector<int> tempSet(size);
            sets = tempSet;
            for (int i = 0; i < size; i++){
                sets[i] = -1;
            } 
        }

        int root(int current){
            int i = current;
            while (sets[i] != -1){
                i = sets[i];
            }
            return i;
        }

        bool isConnected(int s, int t){
            return root(s) == root(t);
        }

        void connect(int s, int t){
            sets[root(s)] = root(t);
        }
};

//Kruskal.hpp
#include "Dijkstra.cpp"
#include "WeightedQuickUnion.cpp"
#include <iostream>
#include <vector>
#include <queue>
#include <limits>
#include <algorithm>
using namespace std;

class Kruskal{
    public:
        std::priority_queue<Edge*, std::vector<Edge*>, Compare> pq;
        std::vector<Edge> mst;
        Kruskal(WeightedGraph wg);
        void kruskal(WeightedGraph wg);
        std::vector<Edge> returnMST();
};

//Kruskal.cpp
#include "Kruskal.hpp"
#include <iostream>
#include <vector>
#include <queue>
#include <limits>
#include <algorithm>
using namespace std;

Kruskal::Kruskal(WeightedGraph wg){
    for (int i = 0; i < wg.V; i++){
        for (Edge j : wg.returnAdjacent(i)){
            pq.push(new Edge(j.source, j.connection, j.weight));
        }
    }
    kruskal(wg);
}

void Kruskal::kruskal(WeightedGraph wg){
    WeightedQuickUnion wqu = WeightedQuickUnion(wg.V);
    while (!pq.empty()){
        Edge* n = pq.top();
        pq.pop();
        if (wqu.isConnected(n->source, n->connection) == false){
            wqu.connect(n->source, n->connection);
            mst.push_back(Edge(n->source, n->connection, n->weight));
        }
    }
}

std::vector<Edge> Kruskal::returnMST(){
    return mst;
}

//main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <functional>
#include <algorithm>
#include "Kruskal.cpp"
using namespace std;

int main(){
    WeightedGraph wg = WeightedGraph(11);
    wg.addEdge(0, 1, 1);
    wg.addEdge(1, 2, 2);
    wg.addEdge(1, 4, 3);
    wg.addEdge(2, 5, 4);
    wg.addEdge(3, 4, 5);
    wg.addEdge(4, 5, 6);
    wg.addEdge(5, 6, 7);
    wg.addEdge(5, 8, 8);
    wg.addEdge(6, 7, 9);
    wg.print();
    std::cout << " --- \n";
    int start = 0;
    int end = 7;

    Dijkstra d = Dijkstra(wg, start);
    std::cout << "Shortest Path From " << start << " to " << end << ": \n";
    for (int i : d.pathTo(end)){
        std::cout << i << " => ";
    }
    std::cout << "\n";
    std::vector<int> distTo = d.getDistTo(end);
    std::cout << "Total Weight: " << distTo[end];
    std::cout << "\n --- \n";

    Kruskal k = Kruskal(wg);
    std::cout << "MST: \n";
    for (Edge i : k.returnMST()){
        std::cout << i.source << " <=> " << i.connection << ": " << i.weight << "\n";
    }
}
